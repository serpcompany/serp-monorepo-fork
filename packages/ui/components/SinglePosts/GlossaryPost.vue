<script setup lang="ts">
  import type { Post } from '@serp/types/types';

  const config = useRuntimeConfig();
  const useAuth = config.public.useAuth;

  const props = defineProps<{
    data: Post;
  }>();

  const data = toRef(props, 'data');

  let comments = data.value.comments || [];
  const commentsData = await usePostComments(data.value.id);
  comments = commentsData.comments;

  // Add safe array of video refs
  const videoRefs = ref<HTMLElement[]>([]);

  // Function to register each video ref as it's mounted
  const setVideoRef = (el: HTMLElement | null, index: number) => {
    if (el) {
      if (!videoRefs.value[index]) {
        videoRefs.value[index] = el;
      }
    }
  };
</script>

<template>
  <div>
    <div class="mt-10 grid grid-cols-1 lg:grid-cols-3">
      <div class="col-span-2 pb-10">
        <h1 class="text-4xl font-bold lg:text-6xl">
          {{ data.keyword || data.title || data.name }}
        </h1>
        <p v-if="data.oneLiner" class="prose dark:prose-invert mt-4">
          {{ data.oneLiner }}
        </p>
      </div>
      <div class="col-span-1 pb-10">
        <SScriptYouTubePlayer
          v-if="data.videoId"
          :ref="(el) => setVideoRef(el as HTMLElement, index)"
          :video-id="data.videoId"
          thumbnail-size="maxresdefault"
        >
          <template #awaitingLoad>
            <div
              class="absolute top-1/2 left-1/2 h-[48px] w-[68px] -translate-x-1/2 -translate-y-1/2 transform"
            >
              <svg height="100%" version="1.1" viewBox="0 0 68 48" width="100%">
                <path
                  d="M66.52,7.74c-0.78-2.93-2.49-5.41-5.42-6.19C55.79,.13,34,0,34,0S12.21,.13,6.9,1.55 C3.97,2.33,2.27,4.81,1.48,7.74C0.06,13.05,0,24,0,24s0.06,10.95,1.48,16.26c0.78,2.93,2.49,5.41,5.42,6.19 C12.21,47.87,34,48,34,48s21.79-0.13,27.1-1.55c2.93-0.78,4.64-3.26,5.42-6.19C67.94,34.95,68,24,68,24S67.94,13.05,66.52,7.74z"
                  fill="#f00"
                />
                <path d="M 45,24 27,14 27,34" fill="#fff" />
              </svg>
            </div>
          </template>
        </SScriptYouTubePlayer>
      </div>

      <USeparator class="col-span-3 my-4">
        <SLogo />
      </USeparator>

      <!-- main section -->
      <div class="prose dark:prose-invert col-span-2">
        <!-- eslint-disable-next-line vue/no-v-html -->
        <article v-html="data.content"></article>
        <div v-if="data.relatedPosts && data.relatedPosts.length === 0">
          <h2>Related Posts</h2>
          <ul>
            <li>Term 1</li>
            <li>Term 2</li>
            <li>Term 3</li>
          </ul>
        </div>

        <!-- Comments Section -->
        <div v-if="useAuth" class="mt-10">
          <h2 class="mb-6 text-3xl font-bold">Comments</h2>
          <CommentsContainer
            :id="data.id"
            module="posts"
            :comments="comments || []"
            class="comments-github-style"
          />
        </div>
      </div>

      <!-- sidebar -->
      <div class="col-span-1"></div>
    </div>
  </div>
</template>

<style>
  .comments-github-style {
    border: 1px solid var(--color-border, #d0d7de);
    border-radius: 6px;
    margin-top: 16px;
  }

  .comments-github-style .innerWrapper {
    padding: 16px;
  }

  .comments-github-style .comment-wrapper {
    border-top: 1px solid var(--color-border, #d0d7de);
    padding-top: 16px;
    margin-top: 16px;
  }

  .comments-github-style .comment-wrapper:first-child {
    border-top: none;
    padding-top: 0;
    margin-top: 0;
  }

  .comments-github-style .wrapper {
    display: flex;
    gap: 16px;
  }

  .comments-github-style .addComment {
    display: flex;
    gap: 16px;
    margin-bottom: 16px;
  }

  .comments-github-style .commentBox {
    flex: 1;
    border: 1px solid var(--color-border, #d0d7de);
    border-radius: 6px;
    padding: 8px;
  }

  .comments-github-style .commentBox textarea {
    width: 100%;
    min-height: 100px;
    padding: 8px;
    border-radius: 3px;
  }

  .comments-github-style .commentBox button {
    margin-top: 8px;
    padding: 5px 16px;
    border-radius: 6px;
    color: white;
  }

  .comments-github-style .name-wrapper {
    display: flex;
    align-items: center;
    gap: 8px;
    margin-bottom: 4px;
  }

  .comments-github-style .time {
    color: #768390;
    font-size: 0.85em;
  }

  .comments-github-style .comment {
    background-color: #f6f8fa;
    border: 1px solid #d0d7de;
    border-radius: 6px;
    padding: 16px;
    margin-bottom: 8px;
  }

  .dark .comments-github-style .comment {
    background-color: #0d1117;
    border-color: #30363d;
  }
</style>
