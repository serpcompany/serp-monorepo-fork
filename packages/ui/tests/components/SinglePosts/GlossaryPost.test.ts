import { mockNuxtImport } from '@nuxt/test-utils/runtime';
import { describe, expect, it } from 'vitest';
import GlossaryPost from '../../../components/SinglePosts/GlossaryPost.vue';
import ComponentRender from '../../componentRender';
import '../../mockUseUserSession';

let runtimeConfig: Record<string, unknown> = { public: { useAuth: true } };
let commentsData_: unknown = { comments: [] };

mockNuxtImport('useHead', () => () => {});
mockNuxtImport('useRuntimeConfig', () => () => runtimeConfig);
mockNuxtImport('usePostComments', () => async () => commentsData_);

describe('SinglePostsGlossaryPost Snapshot', () => {
  const scenarios: [
    string,
    {
      config: Record<string, unknown>;
      props: { data: unknown };
      comments: unknown;
    }
  ][] = [
    [
      'with full post and auth enabled',
      {
        config: { app: { baseURL: '/' }, public: { useAuth: true } },
        props: {
          data: {
            // Using keyword when available (takes precedence over title)
            keyword: 'Nuxt Testing',
            oneLiner: 'A quick guide on snapshot testing.',
            videoId: 'abc123',
            content: '<p>Content of the post</p>',
            // When relatedPosts exists and is empty, a related posts block is rendered.
            relatedPosts: [],
            slug: 'nuxt-testing',
            comments: [
              { author: 'John Doe', comment: 'Great post!', replies: [] }
            ],
            upvotes: ['test@test.com']
          }
        },
        comments: {
          comments: [
            {
              id: 1,
              content: 'Test comment',
              replies: []
            }
          ]
        }
      }
    ],
    [
      'with full post and auth disabled',
      {
        config: { app: { baseURL: '/' }, public: { useAuth: false } },
        props: {
          data: {
            keyword: 'Nuxt Testing',
            oneLiner: 'A quick guide on snapshot testing.',
            videoId: 'abc123',
            content: '<p>Content of the post</p>',
            relatedPosts: [],
            slug: 'nuxt-testing',
            comments: [
              { author: 'John Doe', comment: 'Great post!', replies: [] }
            ],
            upvotes: []
          }
        },
        comments: {
          comments: [
            {
              id: 1,
              content: 'Test comment',
              replies: []
            }
          ]
        }
      }
    ],
    [
      'without video and minimal fields',
      {
        config: { app: { baseURL: '/' }, public: { useAuth: true } },
        props: {
          data: {
            // Omitting keyword, so title will be used for the heading.
            title: 'No Video Post',
            content: '<p>Content without video</p>',
            slug: 'no-video-post',
            // Comments and relatedPosts are omitted or empty.
            comments: [],
            upvotes: []
          }
        },
        comments: { comments: [] }
      }
    ],
    [
      'with non-empty relatedPosts',
      {
        config: { app: { baseURL: '/' }, public: { useAuth: true } },
        props: {
          data: {
            title: 'Related Posts Present',
            oneLiner: 'Post with related posts',
            videoId: 'def456',
            content: '<p>Some content here</p>',
            // Providing non-empty relatedPosts should bypass the related posts block.
            relatedPosts: [{ id: 1, title: 'Related Post' }],
            slug: 'related-posts-post',
            comments: [],
            upvotes: ['test@test.com']
          }
        },
        comments: { comments: [] }
      }
    ]
  ];

  it.each(scenarios)(
    'renders %s correctly',
    async (desc: string, { config, props, comments }) => {
      runtimeConfig = config;
      commentsData_ = comments;

      const html = await ComponentRender(
        `GlossaryPost ${desc}`,
        { props },
        GlossaryPost
      );
      expect(html).toMatchSnapshot();
    }
  );
});
