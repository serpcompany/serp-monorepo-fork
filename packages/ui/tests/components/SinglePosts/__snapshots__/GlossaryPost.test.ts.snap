// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SinglePostsGlossaryPost Snapshot > renders with full post and auth disabled correctly 1`] = `
"<div>
  <div class="mt-10 grid grid-cols-1 lg:grid-cols-3">
    <div class="col-span-2 pb-10">
      <h1 class="text-4xl font-bold lg:text-6xl">Nuxt Testing</h1>
      <p class="prose dark:prose-invert mt-4">A quick guide on snapshot testing.</p>
    </div>
    <div class="col-span-1 pb-10">
      <div aria-busy="false" aria-label="YouTube Player - Loaded" role="application" style="position: relative; background-color: black; max-width: 100%; width: 640px; aspect-ratio: 640 / 360;" thumbnail-size="maxresdefault">
        <div style="width: 100%; height: 100%; position: absolute; top: 0px; left: 0px;"></div><img src="https://i.ytimg.com/vi_webp/abc123/sddefault.webp" alt="" loading="lazy" style="width: 100%; object-fit: cover; height: 100%;">
        <!--v-if-->
      </div>
    </div>
    <div data-orientation="horizontal" role="separator" class="flex items-center align-center text-center w-full flex-row col-span-3 my-4">
      <div class="border-default w-full border-solid border-t"></div>
      <div class="font-medium text-default flex mx-3 whitespace-nowrap">
        <div class="flex items-center space-x-2"><svg width="35pt" viewBox="0 0 1024 1024" xmlns="http://www.w3.org/2000/svg" aria-label="SERP Brand logo">
            <path class="fill-[var(--ui-primary)]" d=" M 127.62 540.68 C 255.64 429.97 383.70 319.31 511.76 208.65 C 639.74 319.23 767.70 429.86 895.69 540.45 C 895.70 631.71 895.73 722.97 895.64 814.23 C 719.72 662.17 543.76 510.14 367.81 358.10 C 398.86 414.80 429.83 471.54 460.92 528.22 C 349.85 623.79 238.69 719.27 127.67 814.91 C 127.66 723.50 127.67 632.09 127.62 540.68 Z"></path>
          </svg></div>
      </div>
      <div class="border-default w-full border-solid border-t"></div>
    </div><!-- main section -->
    <div class="prose dark:prose-invert col-span-2">
      <!-- eslint-disable-next-line vue/no-v-html -->
      <article>
        <p>Content of the post</p>
      </article>
      <div>
        <h2>Related Posts</h2>
        <ul>
          <li>Term 1</li>
          <li>Term 2</li>
          <li>Term 3</li>
        </ul>
      </div><!-- Comments Section -->
      <!--v-if-->
    </div><!-- sidebar -->
    <div class="col-span-1"></div>
  </div>
</div>"
`;

exports[`SinglePostsGlossaryPost Snapshot > renders with full post and auth enabled correctly 1`] = `
"<div>
  <div class="mt-10 grid grid-cols-1 lg:grid-cols-3">
    <div class="col-span-2 pb-10">
      <h1 class="text-4xl font-bold lg:text-6xl">Nuxt Testing</h1>
      <p class="prose dark:prose-invert mt-4">A quick guide on snapshot testing.</p>
    </div>
    <div class="col-span-1 pb-10">
      <div aria-busy="true" aria-label="YouTube Player - Loading" role="application" style="position: relative; background-color: black; max-width: 100%; width: 640px; aspect-ratio: 640 / 360;" thumbnail-size="maxresdefault">
        <div style="width: 100%; height: 100%; position: absolute; top: 0px; left: 0px;"></div><img src="https://i.ytimg.com/vi_webp/abc123/sddefault.webp" alt="" loading="lazy" style="width: 100%; object-fit: cover; height: 100%;">
        <div data-v-e49d796d="" class="loader" styles="[object Object]" aria-label="Loading..." role="status"></div>
      </div>
    </div>
    <div data-orientation="horizontal" role="separator" class="flex items-center align-center text-center w-full flex-row col-span-3 my-4">
      <div class="border-default w-full border-solid border-t"></div>
      <div class="font-medium text-default flex mx-3 whitespace-nowrap">
        <div class="flex items-center space-x-2"><svg width="35pt" viewBox="0 0 1024 1024" xmlns="http://www.w3.org/2000/svg" aria-label="SERP Brand logo">
            <path class="fill-[var(--ui-primary)]" d=" M 127.62 540.68 C 255.64 429.97 383.70 319.31 511.76 208.65 C 639.74 319.23 767.70 429.86 895.69 540.45 C 895.70 631.71 895.73 722.97 895.64 814.23 C 719.72 662.17 543.76 510.14 367.81 358.10 C 398.86 414.80 429.83 471.54 460.92 528.22 C 349.85 623.79 238.69 719.27 127.67 814.91 C 127.66 723.50 127.67 632.09 127.62 540.68 Z"></path>
          </svg></div>
      </div>
      <div class="border-default w-full border-solid border-t"></div>
    </div><!-- main section -->
    <div class="prose dark:prose-invert col-span-2">
      <!-- eslint-disable-next-line vue/no-v-html -->
      <article>
        <p>Content of the post</p>
      </article>
      <div>
        <h2>Related Posts</h2>
        <ul>
          <li>Term 1</li>
          <li>Term 2</li>
          <li>Term 3</li>
        </ul>
      </div><!-- Comments Section -->
      <div class="mt-10">
        <h2 class="mb-6 text-3xl font-bold">Comments</h2>
        <div data-v-e7198dba="" class="comments comments-github-style" style="background: transparent;">
          <div data-v-e7198dba="" class="innerWrapper">
            <div data-v-e7198dba="" class="userName">
              <div data-v-e7198dba="" class="addComment">
                <div data-v-e7198dba="" class="w-12 rounded-t-full">
                  <!--v-if-->
                </div>
                <div data-v-e7198dba="" class="commentBox text-black dark:text-white"><textarea data-v-e7198dba="" name="addComment" class="addComment border-neutral-500 bg-white dark:border-neutral-400 dark:bg-neutral-800" placeholder="Add a comment" spellcheck="true" aria-label="Add comment"></textarea>
                  <!--v-if-->
                  <!--v-if-->
                </div>
              </div>
            </div>
            <transition-group-stub data-v-e7198dba="" name="fade" tag="div" appear="true" persisted="false" css="true">
              <div data-v-d830712d="" data-v-e7198dba="" class="comment-wrapper">
                <div data-v-d830712d="" class="wrapper">
                  <div data-v-d830712d="" class="w-12 rounded-t-full">
                    <!---->
                  </div>
                  <div data-v-d830712d="" class="comment-wrapper">
                    <div data-v-d830712d="" class="name-wrapper">
                      <div data-v-d830712d="" class="name text-(--ui-primary)"></div><span data-v-d830712d="" class="dot">•</span><span data-v-d830712d="" class="time" title="Invalid Date">just now</span>
                      <!--v-if-->
                      <div data-v-d830712d="" class="expand hover:text-(--ui-primary)"><span data-v-d830712d="" title="Hide Comment">−</span></div>
                    </div>
                    <div data-v-d830712d="" class="comment bg-white text-black dark:bg-neutral-800 dark:text-white">Test comment</div>
                    <div data-v-d830712d="" class="comment-box text-black dark:text-white" style="display: none;">
                      <div data-v-d830712d="" class="user-name text-(--ui-primary)"></div><textarea data-v-d830712d="" name="addUpdate" class="add-comment bg-white dark:bg-neutral-800" placeholder="Update message" spellcheck="false" aria-label="Add Comment"></textarea><button data-v-d830712d="" aria-label="Update" class="bg-primary hover:bg-primary-800"><span data-v-d830712d="">Update</span></button>
                      <div data-v-d830712d="" class="remaining-letter bg-primary"><span data-v-d830712d="">1000</span></div>
                    </div>
                    <div data-v-d830712d="" class="reply hover:text-(--ui-primary)" style="display: none;">
                      <div data-v-d830712d="">Never mind</div>
                    </div>
                    <div data-v-d830712d="" class="reply">
                      <div data-v-d830712d="" class="hover:text-(--ui-primary)">Reply</div>
                      <!--v-if-->
                      <!--v-if-->
                    </div>
                    <!--v-if-->
                    <transition-group-stub data-v-d830712d="" name="fade" tag="div" appear="true" persisted="false" css="true"></transition-group-stub>
                    <!--v-if-->
                  </div>
                </div>
              </div>
            </transition-group-stub>
            <!--v-if-->
          </div>
        </div>
      </div>
    </div><!-- sidebar -->
    <div class="col-span-1"></div>
  </div>
</div>"
`;

exports[`SinglePostsGlossaryPost Snapshot > renders with non-empty relatedPosts correctly 1`] = `
"<div>
  <div class="mt-10 grid grid-cols-1 lg:grid-cols-3">
    <div class="col-span-2 pb-10">
      <h1 class="text-4xl font-bold lg:text-6xl">Related Posts Present</h1>
      <p class="prose dark:prose-invert mt-4">Post with related posts</p>
    </div>
    <div class="col-span-1 pb-10">
      <div aria-busy="false" aria-label="YouTube Player - Loaded" role="application" style="position: relative; background-color: black; max-width: 100%; width: 640px; aspect-ratio: 640 / 360;" thumbnail-size="maxresdefault">
        <div style="width: 100%; height: 100%; position: absolute; top: 0px; left: 0px;"></div><img src="https://i.ytimg.com/vi_webp/def456/sddefault.webp" alt="" loading="lazy" style="width: 100%; object-fit: cover; height: 100%;">
        <!--v-if-->
      </div>
    </div>
    <div data-orientation="horizontal" role="separator" class="flex items-center align-center text-center w-full flex-row col-span-3 my-4">
      <div class="border-default w-full border-solid border-t"></div>
      <div class="font-medium text-default flex mx-3 whitespace-nowrap">
        <div class="flex items-center space-x-2"><svg width="35pt" viewBox="0 0 1024 1024" xmlns="http://www.w3.org/2000/svg" aria-label="SERP Brand logo">
            <path class="fill-[var(--ui-primary)]" d=" M 127.62 540.68 C 255.64 429.97 383.70 319.31 511.76 208.65 C 639.74 319.23 767.70 429.86 895.69 540.45 C 895.70 631.71 895.73 722.97 895.64 814.23 C 719.72 662.17 543.76 510.14 367.81 358.10 C 398.86 414.80 429.83 471.54 460.92 528.22 C 349.85 623.79 238.69 719.27 127.67 814.91 C 127.66 723.50 127.67 632.09 127.62 540.68 Z"></path>
          </svg></div>
      </div>
      <div class="border-default w-full border-solid border-t"></div>
    </div><!-- main section -->
    <div class="prose dark:prose-invert col-span-2">
      <!-- eslint-disable-next-line vue/no-v-html -->
      <article>
        <p>Some content here</p>
      </article>
      <!--v-if-->
      <!-- Comments Section -->
      <div class="mt-10">
        <h2 class="mb-6 text-3xl font-bold">Comments</h2>
        <div data-v-e7198dba="" class="comments comments-github-style" style="background: transparent;">
          <div data-v-e7198dba="" class="innerWrapper">
            <div data-v-e7198dba="" class="userName">
              <div data-v-e7198dba="" class="addComment">
                <div data-v-e7198dba="" class="w-12 rounded-t-full">
                  <!--v-if-->
                </div>
                <div data-v-e7198dba="" class="commentBox text-black dark:text-white"><textarea data-v-e7198dba="" name="addComment" class="addComment border-neutral-500 bg-white dark:border-neutral-400 dark:bg-neutral-800" placeholder="Add a comment" spellcheck="true" aria-label="Add comment"></textarea>
                  <!--v-if-->
                  <!--v-if-->
                </div>
              </div>
            </div>
            <transition-group-stub data-v-e7198dba="" name="fade" tag="div" appear="true" persisted="false" css="true"></transition-group-stub>
            <!--v-if-->
          </div>
        </div>
      </div>
    </div><!-- sidebar -->
    <div class="col-span-1"></div>
  </div>
</div>"
`;

exports[`SinglePostsGlossaryPost Snapshot > renders without video and minimal fields correctly 1`] = `
"<div>
  <div class="mt-10 grid grid-cols-1 lg:grid-cols-3">
    <div class="col-span-2 pb-10">
      <h1 class="text-4xl font-bold lg:text-6xl">No Video Post</h1>
      <!--v-if-->
    </div>
    <div class="col-span-1 pb-10">
      <!--v-if-->
    </div>
    <div data-orientation="horizontal" role="separator" class="flex items-center align-center text-center w-full flex-row col-span-3 my-4">
      <div class="border-default w-full border-solid border-t"></div>
      <div class="font-medium text-default flex mx-3 whitespace-nowrap">
        <div class="flex items-center space-x-2"><svg width="35pt" viewBox="0 0 1024 1024" xmlns="http://www.w3.org/2000/svg" aria-label="SERP Brand logo">
            <path class="fill-[var(--ui-primary)]" d=" M 127.62 540.68 C 255.64 429.97 383.70 319.31 511.76 208.65 C 639.74 319.23 767.70 429.86 895.69 540.45 C 895.70 631.71 895.73 722.97 895.64 814.23 C 719.72 662.17 543.76 510.14 367.81 358.10 C 398.86 414.80 429.83 471.54 460.92 528.22 C 349.85 623.79 238.69 719.27 127.67 814.91 C 127.66 723.50 127.67 632.09 127.62 540.68 Z"></path>
          </svg></div>
      </div>
      <div class="border-default w-full border-solid border-t"></div>
    </div><!-- main section -->
    <div class="prose dark:prose-invert col-span-2">
      <!-- eslint-disable-next-line vue/no-v-html -->
      <article>
        <p>Content without video</p>
      </article>
      <!--v-if-->
      <!-- Comments Section -->
      <div class="mt-10">
        <h2 class="mb-6 text-3xl font-bold">Comments</h2>
        <div data-v-e7198dba="" class="comments comments-github-style" style="background: transparent;">
          <div data-v-e7198dba="" class="innerWrapper">
            <div data-v-e7198dba="" class="userName">
              <div data-v-e7198dba="" class="addComment">
                <div data-v-e7198dba="" class="w-12 rounded-t-full">
                  <!--v-if-->
                </div>
                <div data-v-e7198dba="" class="commentBox text-black dark:text-white"><textarea data-v-e7198dba="" name="addComment" class="addComment border-neutral-500 bg-white dark:border-neutral-400 dark:bg-neutral-800" placeholder="Add a comment" spellcheck="true" aria-label="Add comment"></textarea>
                  <!--v-if-->
                  <!--v-if-->
                </div>
              </div>
            </div>
            <transition-group-stub data-v-e7198dba="" name="fade" tag="div" appear="true" persisted="false" css="true"></transition-group-stub>
            <!--v-if-->
          </div>
        </div>
      </div>
    </div><!-- sidebar -->
    <div class="col-span-1"></div>
  </div>
</div>"
`;
